doc/paper/paper.tex

```diff
@@ -0,0 +1,1012 @@
+    rootmap database.
+
+    \subsubsection{Read path}
+    On the read path, the router receives a request as a string in the dp3 query
+    language. The query is parsed and then transformed into an iterator-style
+    execution plan, with a root location retrieved from the tree manager for each
+    topic involved in the query. Scan nodes are instantiated with a timerange and a
+    filter expression and are executed by walking the relevant tree leaves and
+    applying the filter. Leaf nodes are structured as linked lists and sometimes
+    have multiple \q{insert} and \q{delete range} operations, which the
+    scan operator resolves into a set of file read and merge operations.
+
+    Node lookups go through the nodestore, with a read-through byte-capacity
+    bounded LRU cache of inner nodes. If the important inner nodes of the tree are
+    in the cache, then the only reads from storage are those for the leaf data.
+
+    After streaming through the tree of query operators, records are written out
+    to the user in a single topic-multiplexed MCAP data stream.
+
+    \subsection{Public API}
+    The public API of dp3 consists of the following core operations:
+
+    \textbf{Import(database string, producer string, objectID string)} imports an
+    object from storage on behalf of a producer by telling dp3 a location from
+    which to read. The term \q{producer} is intentionally vague: it is up to the
+    user to decide whether to associate producers with \q{devices} or
+    \q{simulation runs} or another internal concept. dp3 supports multiple
+    logical databases within a single instance, and operations must be scoped
+    with a database.
+
+    \textbf{Query(database string, query string)} evaluates the query and
+    returns results in MCAP format.
+
+    \textbf{StatRange(database string, producer string, topic string, start
+    uint64, end uint64, granularity int)} returns a range of field-level
+    statistical aggregations at a granularity at least as granular as the one
+    requested. These statistics are stored in the inner nodes of the tree, so
+    under optimal circumstances StatRange queries should mostly be served from
+    RAM.
+
+    \textbf{Delete(database string, producer string, topic string, start uint64,
+    end uint64)} performs a logical deletion of data between two timestamps.
+    Physically this is an insert of a \q{mask} rather than a true delete.
+    Physical deletion is handled by storage retention policies only.
+
+    \textbf{Truncate(database string, producer string, topic string, timestamp
+    uint64)} perform a logical data truncation at a particular timestamp. This
+    makes data inserted prior to that timestamp unreachable by read requests. To
+    prevent reads from accessing deleted files, it is necessary to truncate
+    tables prior to the time when storage retention policies would delete their
+    objects.
```

Comment:
this doesn't mention the "messages since version" API, and that doesn't mesh well with query. It might be worth a separate item to explain.